README.md
AjaxImageUpload
================================================================================

MODX Snippet/jQuery Script Wrapper for Andrew Valums great file upload script http://valums.com/ajax-upload/

Features:
--------------------------------------------------------------------------------
With this snippet an upload button for uploading multiple files with progress-bar is generated. Works well in FF3.6+, Safari4+, Chrome and falls back to hidden iframe based upload in other browsers, providing good user experience everywhere.

Installation:
--------------------------------------------------------------------------------
1. Upload all files into the new folder *assets/snippets/ajaximageupload*
2. Create a new snippet called AjaxImageUpload with the following snippet code
    `<?php
    return include MODX_BASE_PATH.'assets/snippets/ajaximageupload/ajaximageupload.snippet.php';
    ?>`

Usage
--------------------------------------------------------------------------------

- Non Ajax Page: 
```
[!AjaxImageUpload? &mode=`form` &ajaxId=`779` &uid=`uniqueid` &language=`german` &allowedExtensions=`jpg,jpeg,png,gif` &thumbX=`75` &thumbY=`75`!]
```
- AJAX page: 
```
[!AjaxImageUpload? &mode=`ajax` &language=`german` &allowedExtensions=`jpg,jpeg,png,gif` &maxFilesizeMb=`2` &thumbX=`100` &thumbY=`100`!]
```

Parameters
--------------------------------------------------------------------------------

Property | Description | Default
---- | ----------- | -------
mode | Snippet mode | form
formUid | Unique form Id |  md5 of MODX 'site_url' setting
language | Snippet language | english
allowedExtensions | Allowed file extensions for upload | jpg,jpeg,png,gif
maxFilesizeMb | Maximum size for one file to upload | 8
maxFiles | Maximum count of files to upload | 3
thumbX | horizontal size of generated thumb | 100 
thumbY | vertical size of generated thumb | 100 
ajaxId | id of the document with the ajax snippet call | 0
addJquery | add jQuery script in head | 1
addJscript | add the snippet javascript and the fileuploader script in head | 1
addCss | add the snippet css in head | 1


Notes:
--------------------------------------------------------------------------------
1. The uploaded images will be saved with an unique filename in `assets/cache/ajaximageupload`. This folder should be cleaned from time to time.
2. The properties of the uploaded images will be saved in `$_SESSION['AjaxImageUpload'][$uid][$id]` as `array('originalName', 'uniqueName', 'thumbName', 'path')` and should be moved from there during the other form process. The parameter `$uid` should be set on non ajax page with the parameter `formUid`, `$id` is the number of the image.
3. The parameter `formUid` could be generated and set for each upload button (to separate multiple upload queues) and maybe each pageview (to separate uploads for the same queue and session but i.e. in different browser tabs).
4. Different upload buttons could use different ajax 'documents' for allowing different file types, thumb sizes etc.

HISTORY.md
History
================================================================================

- 1.1
    - install set for http://extras.evolution-cms.com/
- 1.0
	- Initial public release


AjaxImageUpload.js
(function($) {
	$.fn.ajaxFileUpload = function(options) {
		var settings = $.extend({
			'debug' : false,
			'uploadAction' : '',
			'uid' : '',
			'uploadTemplate' : '<div class="qq-uploader"><div class="qq-upload-drop-area"><span>Drop files here to upload</span></div><div class="qq-upload-button">Upload a file</div><ul class="qq-upload-list"></ul></div>',
			'fileTemplate' : '<li><span class="qq-upload-file"></span><span class="qq-upload-spinner"></span><span class="qq-upload-size"></span><a class="qq-upload-cancel" href="#">Cancel</a><span class="qq-upload-failed-text">Failed</span></li>',
			'deleteTemplate' : '<div class="delete-button">Delete image</div>',
			'clearTemplate' : '<div class="qq-clear-button">Delete all images</div>',
			'thumbX' : '100px',
			'thumbY' : '100px',
			'allowedExtensions' : [],
			'sizeLimit' : 0,
			'messages' : {
				'typeError' : "{file} has invalid extension. Only {extensions} are allowed.",
				'sizeError' : "{file} is too large, maximum file size is {sizeLimit}.",
				'minSizeError' : "{file} is too small, minimum file size is {minSizeLimit}.",
				'emptyError' : "{file} is empty, please select files again without it.",
				'onLeave' : "The files are being uploaded, if you leave now the upload will be cancelled."
			}
		}, options);

		return this.each(function() {
			var element = $(this);
			var imageList = element.find('.file-uploader-images');
			var uploader = new qq.FileUploader({
				'element' : element.find('.file-uploader-buttons')[0],
				'action' : settings.uploadAction,
				'params' : {
					'uid' : settings.uid
				},
				'template' : settings.uploadTemplate,
				'fileTemplate' : settings.fileTemplate,
				'allowedExtensions' : settings.allowedExtensions,
				'sizeLimit' : settings.sizeLimit,
				'messages' : settings.messages,
				'onComplete' : function(id, fileName, uploadAnswer) {
					var fileid = uploadAnswer.fileid;
					if(uploadAnswer.success) {
						var imageWrap = $('<div>').addClass('image-wrap').addClass('image' + fileid);
						var deleteButton = $(settings.deleteTemplate).click(function() {
							$.get(settings.uploadAction, {
								'delete' : fileid
							}, function(deleteAnswer) {
								if(deleteAnswer.success) {
									imageWrap.fadeOut(function() {
										$(this).remove();
									});
									if(settings.debug) {
										alert(JSON.stringify(deleteAnswer) + '\nImage ' + fileid + ' deleted.');
									}
								}
							}, 'json');
						});
						var image = $('<img>').attr({
							src : uploadAnswer.filename
						}).css({
							'width' : settings.thumbX,
							'height' : settings.thumbY,
							'position' : 'relative'
						}).after(deleteButton);

						imageList.append(imageWrap.append(image));
						element.find('.qq-upload-list li').eq(id).hide();
						if(settings.debug) {
							alert('ID:' + id + '\nResponse:' + JSON.stringify(uploadAnswer));
						}
					}
				}
			});
			var clearButton = $(settings.clearTemplate).click(function() {
				$.get(settings.uploadAction, {
					'delete' : 'all',
					'uid' : settings.uid
				}, function(clearAnswer) {
					if(clearAnswer.success) {
						imageList.empty();
						element.find('.qq-upload-list li').remove();
						if(settings.debug) {
							alert(JSON.stringify(deleteAnswer) + '\nAll Images deleted.');
						}
					}
				}, 'json');
			});
			imageList.find('.image-wrap .delete-button').click(function() {
				var imageWrap = $(this).parent();
				var imageClasses = imageWrap.attr("class").split(/\s+/);
				for( i = 0; i < imageClasses.length; i++) {
					var fileid = imageClasses[i].split("image")[1];
					$.get(settings.uploadAction, {
						'delete' : parseInt(fileid),
						'uid' : settings.uid
					}, function(deleteAnswer) {
						if(deleteAnswer.success) {
							imageWrap.fadeOut(function() {
								$(this).remove();
							});
							if(settings.debug) {
								alert(JSON.stringify(deleteAnswer) + '\nImage ' + fileid + ' deleted.');
							}
						}
					}, 'json');
				}
			});
			element.find('.qq-upload-button').after(clearButton);
		});
	};
})(jQuery);


AjaxImageUpload.snippet.php
<?php
define('AIU_BASE_PATH', MODX_BASE_PATH . 'assets/snippets/AjaxImageUpload/');
define('AIU_CACHE_PATH', MODX_BASE_PATH . 'assets/cache/AjaxImageUpload/');

include AIU_BASE_PATH . 'includes/fileuploader/fileuploader.class.php';
include AIU_BASE_PATH . 'includes/PhpThumbFactory/ThumbLib.inc.php';

/* * ************************ */
/* Set/Read Snippet Params */
/* * ************************ */

// default: &language=`english` &allowedExtensions=`jpg,jpeg,png,gif` &maxFilesizeMb=`8` &uid=`site-specific` &maxFiles=`3` &thumbX=`100` &thumbY=`100` &mode=`form` &ajaxId=`0`

$language = isset($language) ? $language : 'english';
// comma separated list of valid extensions
$allowedExtensions = isset($allowedExtensions) ? $allowedExtensions : 'jpg,jpeg,png,gif';
$maxFilesizeMb = isset($maxFilesizeMb) ? intval($maxFilesizeMb) : 8;
$formUid = isset($uid) ? $uid : md5($modx->config['site_url']);
$maxFiles = isset($maxFiles) ? $maxFiles : '3';
$thumbX = isset($thumbX) ? $thumbX : '100';
$thumbY = isset($thumbY) ? $thumbY : '100';
$mode = isset($mode) ? $mode : 'form';
$ajaxId = isset($ajaxId) ? intval($ajaxId) : 0;
$addJquery = isset($addJquery) ? intval($addJquery) : 1;
$addJscript = isset($addJscript) ? intval($addJquery) : 1;
$addCss = isset($addCss) ? intval($addJquery) : 1;

function includeFileName($name, $type = 'config', $defaultName = 'default', $fileType = '.inc.php') {
	$folder = (substr($type, -1) != 'y') ? $type . 's/' : substr($folder, 0, -1) . 'ies/';
	$allowedConfigs = glob(AIU_BASE_PATH . $folder . '*.' . $type . $fileType);
	foreach ($allowedConfigs as $config) {
		$configs[] = preg_replace('=.*/' . $folder . '([^.]*).' . $type . $fileType . '=', '$1', $config);
	}
	if (in_array($name, $configs)) {
		return AIU_BASE_PATH . $folder . $name . '.' . $type . $fileType;
	} else {
		if (file_exists(AIU_BASE_PATH . $folder . $defaultName . '.' . $type . $fileType)) {
			return AIU_BASE_PATH . $folder . $defaultName . '.' . $type . $fileType;
		} else {
			$modx->messageQuit('Default AjaxImageUpload ' . $type . ' file "' . AIU_BASE_PATH . $folder . $defaultName . '.' . $type . '.inc.php" not found. Did you upload all snippet files?');
		}
	}
}

if (!file_exists(MODX_BASE_PATH . 'assets/cache/AjaxImageUpload')) {
	mkdir(MODX_BASE_PATH . 'assets/cache/AjaxImageUpload', 0755);
}

include (includeFileName($language, 'language', 'english'));
$allowedExtensions = explode(',', $allowedExtensions);
$sizeLimit = intval($maxFilesizeMb) * 1024 * 1024;
switch ($mode) {
	// the AJAX part of the snippet
	case 'ajax' : {
			// delete uploaded images
			if (isset($_GET['delete'])) {
				$result = array();
				$formUid = (isset($_GET['uid'])) ? htmlentities(trim($_GET['uid']), ENT_NOQUOTES) : $formUid;
				if (strtolower($_GET['delete']) == 'all') {
					// delete all uploaded files/thumbs & clean session
					if (is_array($_SESSION['AjaxImageUpload'][$formUid])) {
						foreach ($_SESSION['AjaxImageUpload'][$formUid] as $key => $fileInfo) {
							if (file_exists($fileInfo['path'] . $fileInfo['uniqueName'])) {
								unlink($fileInfo['path'] . $fileInfo['uniqueName']);
							}
							if (isset($fileInfo['thumbName']) && file_exists($fileInfo['path'] . $fileInfo['thumbName'])) {
								unlink($fileInfo['path'] . $fileInfo['thumbName']);
							}
						}
					}
					$_SESSION['AjaxImageUpload'][$formUid] = array();
					$result['success'] = TRUE;
				} else {
					// delete one uploaded file/thumb & remove session entry
					$fileId = intval($_GET['delete']);
					if (isset($_SESSION['AjaxImageUpload'][$formUid][$fileId])) {
						$fileInfo = $_SESSION['AjaxImageUpload'][$formUid][$fileId];
						if (file_exists($fileInfo['path'] . $fileInfo['uniqueName'])) {
							unlink($fileInfo['path'] . $fileInfo['uniqueName']);
						}
						if (isset($fileInfo['thumbName']) && file_exists($fileInfo['path'] . $fileInfo['thumbName'])) {
							unlink($fileInfo['path'] . $fileInfo['thumbName']);
						}
						unset($_SESSION['AjaxImageUpload'][$formUid][$fileId]);
						$result['success'] = TRUE;
					} else {
						$result['error'] = sprintf($language['notFound'], $maxFiles);
					}
				}
			} else {
				// upload the image(s)
				$uploader = new qqFileUploader($allowedExtensions, $sizeLimit);
				$formUid = (isset($_GET['uid'])) ? htmlentities(trim($_GET['uid']), ENT_NOQUOTES) : $formUid;
				// to pass data through iframe you will need to encode all html tags
				$result = $uploader->handleUpload(AIU_CACHE_PATH, TRUE, $language);
				// file successful uploaded
				if ($result['success']) {
					$originalName = $uploader->filename . '.' . $uploader->extension;
					$path = $uploader->path;
					// check if count of uploaded files are below max file count
					if (count($_SESSION['AjaxImageUpload'][$formUid]) < $maxFiles) {
						// create unique filename and unique thumbname
						$uniqueName = md5($uploader->filename . time()) . '.' . $uploader->extension;
						$thumbName = md5($uploader->filename . time() . '.thumb') . '.' . $uploader->extension;
						// generate thumbname
						$thumb = PhpThumbFactory::create($path . $originalName);
						$thumb->adaptiveResize($thumbX, $thumbY);
						$thumb->save($path . $thumbName);
						rename($path . $originalName, $path . $uniqueName);
						// fill session
						$_SESSION['AjaxImageUpload'][$formUid][] = array('originalName' => $originalName, 'uniqueName' => $uniqueName, 'thumbName' => $thumbName, 'path' => $path);
						// prepare returned values (filename & fileid)
						$result['filename'] = str_replace(MODX_BASE_PATH, $modx->config['site_url'], $path . $thumbName);
						$result['fileid'] = end(array_keys($_SESSION['AjaxImageUpload'][$formUid]));
					} else {
						unset($result['success']);
						// error message
						$result['error'] = sprintf($language['maxFiles'], $maxFiles);
						// delete uploaded file
						unlink($path . $originalName);
					}
				}
			}
			echo htmlspecialchars(json_encode($result), ENT_NOQUOTES);
			exit;
		}
	case 'form' : {
			if ($ajaxId) {
				if ($addJquery) {
					$modx->regClientStartupScript('http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js');
				}
				if ($addCss) {
					$modx->regClientCSS('assets/snippets/AjaxImageUpload/AjaxImageUpload.css');
				}
				if ($addJscript) {
					$modx->regClientStartupScript('assets/snippets/AjaxImageUpload/includes/fileuploader/fileuploader.js');
					$modx->regClientStartupScript('assets/snippets/AjaxImageUpload/AjaxImageUpload.js');
				}
				$scriptSettings = file_get_contents(includeFileName('script' . ucfirst($language), 'template', 'script', '.html'));
				$placeholder = array();
				$placeholder['ajaxId'] = $modx->makeUrl($ajaxId);
				$placeholder['dropArea'] = $language['dropArea'];
				$placeholder['uploadButton'] = $language['uploadButton'];
				$placeholder['clearButton'] = $language['clearButton'];
				$placeholder['cancel'] = $language['cancel'];
				$placeholder['failed'] = $language['failed'];
				$placeholder['thumbX'] = $thumbX;
				$placeholder['thumbY'] = $thumbY;
				$placeholder['allowedExtensions'] = (count($allowedExtensions)) ? "'" . implode("', '", $allowedExtensions) . "'" : '[]';
				$placeholder['sizeLimit'] = $sizeLimit;
				$placeholder['uid'] = $formUid;
				$placeholder['typeError'] = $language['typeError'];
				$placeholder['sizeError'] = $language['sizeError'];
				$placeholder['minSizeError'] = $language['minSizeError'];
				$placeholder['emptyError'] = $language['emptyError'];
				$placeholder['onLeave'] = $language['onLeave'];
				foreach ($placeholder as $key => $value) {
					$scriptSettings = str_replace('[+' . $key . '+]', $value, $scriptSettings);
				}
				$modx->regClientStartupScript($scriptSettings);
				$output = file_get_contents(includeFileName('uploadSection' . ucfirst($language), 'template', 'uploadSection', '.html'));
				$imageTpl = file_get_contents(includeFileName('image' . ucfirst($language), 'template', 'image', '.html'));
				$imageList = array();
				$placeholder = array();
				$placeholder['thumbX'] = $thumbX;
				$placeholder['thumbY'] = $thumbY;
				$placeholder['deleteButton'] = $language['deleteButton'];
				if (is_array($_SESSION['AjaxImageUpload'][$formUid])) {
					foreach ($_SESSION['AjaxImageUpload'][$formUid] as $id => &$fileInfo) {
						$placeholder['id'] = $id;
						if (file_exists($fileInfo['path'] . $fileInfo['uniqueName'])) {
							if (isset($fileInfo['thumbName'])) {
								$placeholder['thumbName'] = str_replace(MODX_BASE_PATH, $modx->config['site_url'], $fileInfo['path'] . $fileInfo['thumbName']);
							} else {
								$path_info = pathinfo($fileInfo['uniqueName']);
								$thumbName = md5($path_info['basename'] . time() . '.thumb') . '.' . $path_info['extension'];
								$thumb = PhpThumbFactory::create($fileInfo['path'] . $fileInfo['uniqueName']);
								$thumb->adaptiveResize($thumbX, $thumbY);
								$thumb->save($fileInfo['path'] . $thumbName);
								$fileInfo['thumbName'] = $thumbName;
								$placeholder['thumbName'] = str_replace(MODX_BASE_PATH, $modx->config['site_url'], $fileInfo['path'] . $thumbName);
							}
							$imageElement = $imageTpl;
							foreach ($placeholder as $key => $value) {
								$imageElement = str_replace('[+' . $key . '+]', $value, $imageElement);
							}
							$imageList[] = $imageElement;
						} else {
							unset($fileInfo);
						}
					}
				}
				$output = str_replace('[+images+]', implode("\r\n", $imageList), $output);
				$output = str_replace('[+uid+]', $formUid, $output);
				return $output;
				break;
			} else {
				return;
			}
		}
}
?>

includes/PhpThumbFactory/PhpThumb.inc.php
<?php
/**
 * PhpThumbGxd Library Definition File
 * 
 * This file contains the definitions for the PhpThumbGxd class.
 * 
 * PHP Version 5 with GD 2.0+
 * PhpThumbGxd : PHP Thumb Library <http://phpthumb.gxdlabs.com>
 * Copyright (c) 2009, Ian Selby/Gen X Design
 * 
 * Author(s): Ian Selby <ian@gen-x-design.com>
 * 
 * Licensed under the MIT License
 * Redistributions of files must retain the above copyright notice.
 * 
 * @author Ian Selby <ian@gen-x-design.com>
 * @copyright Copyright (c) 2009 Gen X Design
 * @link http://phpthumb.gxdlabs.com
 * @license http://www.opensource.org/licenses/mit-license.php The MIT License
 * @version 3.0
 * @package PhpThumbGxd
 * @filesource
 */



/**
 * PhpThumbGxd Object
 * 
 * This singleton object is essentially a function library that helps with core validation 
 * and loading of the core classes and plugins.  There isn't really any need to access it directly, 
 * unless you're developing a plugin and need to take advantage of any of the functionality contained 
 * within.
 * 
 * If you're not familiar with singleton patterns, here's how you get an instance of this class (since you 
 * can't create one via the new keyword):
 * <code>$pt = PhpThumbGxd::getInstance();</code>
 * 
 * It's that simple!  Outside of that, there's no need to modify anything within this class, unless you're doing 
 * some crazy customization... then knock yourself out! :)
 * 
 * @package PhpThumbGxd
 * @subpackage Core
 */
class PhpThumbGxd
{
	/**
	 * Instance of self
	 * 
	 * @var object PhpThumbGxd
	 */
	protected static $_instance;
	/**
	 * The plugin registry
	 * 
	 * This is where all plugins to be loaded are stored.  Data about the plugin is 
	 * provided, and currently consists of:
	 *  - loaded: true/false
	 *  - implementation: gd/imagick/both
	 * 
	 * @var array
	 */
	protected $_registry;
	/**
	 * What implementations are available
	 * 
	 * This stores what implementations are available based on the loaded 
	 * extensions in PHP, NOT whether or not the class files are present.
	 * 
	 * @var array
	 */
	protected $_implementations;
	
	/**
	 * Returns an instance of self
	 * 
	 * This is the usual singleton function that returns / instantiates the object
	 * 
	 * @return PhpThumbGxd
	 */
	public static function getInstance ()
	{
		if(!(self::$_instance instanceof self))
		{
			self::$_instance = new self();
		}

		return self::$_instance;
	}
	
	/**
	 * Class constructor
	 * 
	 * Initializes all the variables, and does some preliminary validation / checking of stuff
	 * 
	 */
	private function __construct ()
	{
		$this->_registry		= array();
		$this->_implementations	= array('gd' => false, 'imagick' => false);
		
		$this->getImplementations();
	}
	
	/**
	 * Finds out what implementations are available
	 * 
	 * This function loops over $this->_implementations and validates that the required extensions are loaded.
	 * 
	 * I had planned on attempting to load them dynamically via dl(), but that would provide more overhead than I 
	 * was comfortable with (and would probably fail 99% of the time anyway)
	 * 
	 */
	private function getImplementations ()
	{
		foreach($this->_implementations as $extension => $loaded)
		{
			if($loaded)
			{
				continue;
			}
			
			if(extension_loaded($extension))
			{
				$this->_implementations[$extension] = true;
			}
		}
	}
	
	/**
	 * Returns whether or not $implementation is valid (available)
	 * 
	 * If 'all' is passed, true is only returned if ALL implementations are available.
	 * 
	 * You can also pass 'n/a', which always returns true
	 * 
	 * @return bool 
	 * @param string $implementation
	 */
	public function isValidImplementation ($implementation)
	{
		if ($implementation == 'n/a')
		{
			return true;
		}
		
		if ($implementation == 'all')
		{
			foreach ($this->_implementations as $imp => $value)
			{
				if ($value == false)
				{
					return false;
				}
			}
			
			return true;
		}
		
		if (array_key_exists($implementation, $this->_implementations))
		{
			return $this->_implementations[$implementation];
		}
		
		return false;
	}
	
	/**
	 * Registers a plugin in the registry
	 * 
	 * Adds a plugin to the registry if it isn't already loaded, and if the provided 
	 * implementation is valid.  Note that you can pass the following special keywords 
	 * for implementation:
	 *  - all - Requires that all implementations be available
	 *  - n/a - Doesn't require any implementation
	 *  
	 * When a plugin is added to the registry, it's added as a key on $this->_registry with the value 
	 * being an array containing the following keys:
	 *  - loaded - whether or not the plugin has been "loaded" into the core class
	 *  - implementation - what implementation this plugin is valid for
	 * 
	 * @return bool
	 * @param string $pluginName
	 * @param string $implementation
	 */
	public function registerPlugin ($pluginName, $implementation)
	{
		if (!array_key_exists($pluginName, $this->_registry) && $this->isValidImplementation($implementation))
		{
			$this->_registry[$pluginName] = array('loaded' => false, 'implementation' => $implementation);
			return true;
		}
		
		return false;
	}
	
	/**
	 * Loads all the plugins in $pluginPath
	 * 
	 * All this function does is include all files inside the $pluginPath directory.  The plugins themselves 
	 * will not be added to the registry unless you've properly added the code to do so inside your plugin file.
	 * 
	 * @param string $pluginPath
	 */
	public function loadPlugins ($pluginPath)
	{
		// strip the trailing slash if present
		if (substr($pluginPath, strlen($pluginPath) - 1, 1) == '/')
		{
			$pluginPath = substr($pluginPath, 0, strlen($pluginPath) - 1);
		}
		
		if ($handle = opendir($pluginPath))
		{
			while (false !== ($file = readdir($handle)))
			{
				if ($file == '.' || $file == '..' || $file == '.svn')
				{
					continue;
				}
				
				include_once($pluginPath . '/' . $file);
			}
		}
	}
	
	/**
	 * Returns the plugin registry for the supplied implementation
	 * 
	 * @return array
	 * @param string $implementation
	 */
	public function getPluginRegistry ($implementation)
	{
		$returnArray = array();
		
		foreach ($this->_registry as $plugin => $meta)
		{
			if ($meta['implementation'] == 'n/a' || $meta['implementation'] == $implementation)
			{
				$returnArray[$plugin] = $meta;
			}
		}
		
		return $returnArray;
	}
}


assets/snippets/ajaximageupload/ajaximageupload.snippet.php
<?php
/*
 * AjaxImageUpload
 *
 * MODX Snippet/jQuery Script Wrapper for Andrew Valums great file upload script
 *
 * License GPL
 * Version 1.1 (05. November 2013)
 * Author: Jako
 */

define('AIU_PATH', 'assets/snippets/ajaximageupload/');
define('AIU_BASE_PATH', MODX_BASE_PATH . AIU_PATH);
define('AIU_CACHE_PATH', 'assets/cache/ajaximageupload/');
define('AIU_BASE_CACHE_PATH', MODX_BASE_PATH . AIU_CACHE_PATH);

include AIU_BASE_PATH . 'includes/fileuploader/fileuploader.class.php';
include AIU_BASE_PATH . 'includes/PhpThumbFactory/ThumbLib.inc.php';

// Set/Read Snippet Params
// default: &language=`english` &allowedExtensions=`jpg,jpeg,png,gif` &maxFilesizeMb=`8` &uid=`site-specific` &maxFiles=`3` &thumbX=`100` &thumbY=`100` &mode=`form` &ajaxId=`0`

$language = isset($language) ? $language : 'english';
// comma separated list of valid extensions
$allowedExtensions = isset($allowedExtensions) ? $allowedExtensions : 'jpg,jpeg,png,gif';
$maxFilesizeMb = isset($maxFilesizeMb) ? intval($maxFilesizeMb) : 8;
$formUid = isset($uid) ? $uid : md5($modx->config['site_url']);
$maxFiles = isset($maxFiles) ? $maxFiles : '3';
$thumbX = isset($thumbX) ? $thumbX : '100';
$thumbY = isset($thumbY) ? $thumbY : '100';
$mode = isset($mode) ? $mode : 'form';
$ajaxId = isset($ajaxId) ? intval($ajaxId) : 0;
$addJquery = isset($addJquery) ? intval($addJquery) : 1;
$addJscript = isset($addJscript) ? intval($addJquery) : 1;
$addCss = isset($addCss) ? intval($addJquery) : 1;

function includeFileName($name, $type = 'config', $defaultName = 'default', $fileType = '.inc.php') {
	$folder = (substr($type, -1) != 'y') ? $type . 's/' : substr($folder, 0, -1) . 'ies/';
	$allowedConfigs = glob(AIU_BASE_PATH . $folder . '*.' . $type . $fileType);
	foreach ($allowedConfigs as $config) {
		$configs[] = preg_replace('=.*/' . $folder . '([^.]*).' . $type . $fileType . '=', '$1', $config);
	}
	if (in_array($name, $configs)) {
		return AIU_BASE_PATH . $folder . $name . '.' . $type . $fileType;
	} else {
		if (file_exists(AIU_BASE_PATH . $folder . $defaultName . '.' . $type . $fileType)) {
			return AIU_BASE_PATH . $folder . $defaultName . '.' . $type . $fileType;
		} else {
			$modx->messageQuit('Default AjaxImageUpload ' . $type . ' file "' . AIU_BASE_PATH . $folder . $defaultName . '.' . $type . '.inc.php" not found. Did you upload all snippet files?');
		}
	}
}

if (!file_exists(AIU_BASE_CACHE_PATH)) {
	mkdir(AIU_BASE_CACHE_PATH, 0755);
}

include (includeFileName($language, 'language', 'english'));
$allowedExtensions = explode(',', $allowedExtensions);
$sizeLimit = intval($maxFilesizeMb) * 1024 * 1024;
switch ($mode) {
	// the AJAX part of the snippet
	case 'ajax' : {
			// delete uploaded images
			if (isset($_GET['delete'])) {
				$result = array();
				$formUid = (isset($_GET['uid'])) ? htmlentities(trim($_GET['uid']), ENT_NOQUOTES) : $formUid;
				if (strtolower($_GET['delete']) == 'all') {
					// delete all uploaded files/thumbs & clean session
					if (is_array($_SESSION['AjaxImageUpload'][$formUid])) {
						foreach ($_SESSION['AjaxImageUpload'][$formUid] as $key => $fileInfo) {
							if (file_exists($fileInfo['path'] . $fileInfo['uniqueName'])) {
								unlink($fileInfo['path'] . $fileInfo['uniqueName']);
							}
							if (isset($fileInfo['thumbName']) && file_exists($fileInfo['path'] . $fileInfo['thumbName'])) {
								unlink($fileInfo['path'] . $fileInfo['thumbName']);
							}
						}
					}
					$_SESSION['AjaxImageUpload'][$formUid] = array();
					$result['success'] = TRUE;
				} else {
					// delete one uploaded file/thumb & remove session entry
					$fileId = intval($_GET['delete']);
					if (isset($_SESSION['AjaxImageUpload'][$formUid][$fileId])) {
						$fileInfo = $_SESSION['AjaxImageUpload'][$formUid][$fileId];
						if (file_exists($fileInfo['path'] . $fileInfo['uniqueName'])) {
							unlink($fileInfo['path'] . $fileInfo['uniqueName']);
						}
						if (isset($fileInfo['thumbName']) && file_exists($fileInfo['path'] . $fileInfo['thumbName'])) {
							unlink($fileInfo['path'] . $fileInfo['thumbName']);
						}
						unset($_SESSION['AjaxImageUpload'][$formUid][$fileId]);
						$result['success'] = TRUE;
					} else {
						$result['error'] = $language['notFound'];
					}
				}
			} else {
				// upload the image(s)
				$uploader = new qqFileUploader($allowedExtensions, $sizeLimit);
				$formUid = (isset($_GET['uid'])) ? htmlentities(trim($_GET['uid']), ENT_NOQUOTES) : $formUid;
				// to pass data through iframe you will need to encode all html tags
				$result = $uploader->handleUpload(AIU_BASE_CACHE_PATH, TRUE, $language);
				// file successful uploaded
				if ($result['success']) {
					$originalName = $uploader->filename . '.' . $uploader->extension;
					$path = $uploader->path;
					// check if count of uploaded files are below max file count
					if (count($_SESSION['AjaxImageUpload'][$formUid]) < $maxFiles) {
						// create unique filename and unique thumbname
						$uniqueName = md5($uploader->filename . time()) . '.' . $uploader->extension;
						$thumbName = md5($uploader->filename . time() . '.thumb') . '.' . $uploader->extension;
						// generate thumbname
						$thumb = PhpThumbFactory::create($path . $originalName);
						$thumb->adaptiveResize($thumbX, $thumbY);
						$thumb->save($path . $thumbName);
						rename($path . $originalName, $path . $uniqueName);
						// fill session
						$_SESSION['AjaxImageUpload'][$formUid][] = array('originalName' => $originalName, 'uniqueName' => $uniqueName, 'thumbName' => $thumbName, 'path' => $path);
						// prepare returned values (filename & fileid)
						$result['filename'] = str_replace(MODX_BASE_PATH, $modx->config['site_url'], $path . $thumbName);
						$result['fileid'] = end(array_keys($_SESSION['AjaxImageUpload'][$formUid]));
					} else {
						unset($result['success']);
						// error message
						$result['error'] = sprintf($language['maxFiles'], $maxFiles);
						// delete uploaded file
						unlink($path . $originalName);
					}
				}
			}
			echo htmlspecialchars(json_encode($result), ENT_NOQUOTES);
			exit;
		}
	case 'form' : {
			if ($ajaxId) {
				if ($addJquery) {
					$modx->regClientStartupScript('http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js');
				}
				if ($addCss) {
					$modx->regClientCSS(AIU_PATH . 'ajaximageupload.css');
				}
				if ($addJscript) {
					$modx->regClientStartupScript(AIU_PATH . 'includes/fileuploader/fileuploader.js');
					$modx->regClientStartupScript(AIU_PATH . 'ajaximageupload.js');
				}
				$scriptSettings = file_get_contents(includeFileName('script' . ucfirst($language), 'template', 'script', '.html'));
				$placeholder = array();
				$placeholder['ajaxId'] = $modx->makeUrl($ajaxId);
				$placeholder['dropArea'] = $language['dropArea'];
				$placeholder['uploadButton'] = $language['uploadButton'];
				$placeholder['clearButton'] = $language['clearButton'];
				$placeholder['cancel'] = $language['cancel'];
				$placeholder['failed'] = $language['failed'];
				$placeholder['thumbX'] = $thumbX;
				$placeholder['thumbY'] = $thumbY;
				$placeholder['allowedExtensions'] = (count($allowedExtensions)) ? "'" . implode("', '", $allowedExtensions) . "'" : '[]';
				$placeholder['sizeLimit'] = $sizeLimit;
				$placeholder['uid'] = $formUid;
				$placeholder['typeError'] = $language['typeError'];
				$placeholder['sizeError'] = $language['sizeError'];
				$placeholder['minSizeError'] = $language['minSizeError'];
				$placeholder['emptyError'] = $language['emptyError'];
				$placeholder['onLeave'] = $language['onLeave'];
				foreach ($placeholder as $key => $value) {
					$scriptSettings = str_replace('[+' . $key . '+]', $value, $scriptSettings);
				}
				$modx->regClientStartupScript($scriptSettings);
				$output = file_get_contents(includeFileName('uploadSection' . ucfirst($language), 'template', 'uploadSection', '.html'));
				$imageTpl = file_get_contents(includeFileName('image' . ucfirst($language), 'template', 'image', '.html'));
				$imageList = array();
				$placeholder = array();
				$placeholder['thumbX'] = $thumbX;
				$placeholder['thumbY'] = $thumbY;
				$placeholder['deleteButton'] = $language['deleteButton'];
				if (is_array($_SESSION['AjaxImageUpload'][$formUid])) {
					foreach ($_SESSION['AjaxImageUpload'][$formUid] as $id => &$fileInfo) {
						$placeholder['id'] = $id;
						if (file_exists($fileInfo['path'] . $fileInfo['uniqueName'])) {
							if (isset($fileInfo['thumbName'])) {
								$placeholder['thumbName'] = str_replace(MODX_BASE_PATH, $modx->config['site_url'], $fileInfo['path'] . $fileInfo['thumbName']);
							} else {
								$path_info = pathinfo($fileInfo['uniqueName']);
								$thumbName = md5($path_info['basename'] . time() . '.thumb') . '.' . $path_info['extension'];
								$thumb = PhpThumbFactory::create($fileInfo['path'] . $fileInfo['uniqueName']);
								$thumb->adaptiveResize($thumbX, $thumbY);
								$thumb->save($fileInfo['path'] . $thumbName);
								$fileInfo['thumbName'] = $thumbName;
								$placeholder['thumbName'] = str_replace(MODX_BASE_PATH, $modx->config['site_url'], $fileInfo['path'] . $thumbName);
							}
							$imageElement = $imageTpl;
							foreach ($placeholder as $key => $value) {
								$imageElement = str_replace('[+' . $key . '+]', $value, $imageElement);
							}
							$imageList[] = $imageElement;
						} else {
							unset($fileInfo);
						}
					}
				}
				$output = str_replace('[+images+]', implode("\r\n", $imageList), $output);
				$output = str_replace('[+uid+]', $formUid, $output);
				return $output;
				break;
			} else {
				return;
			}
		}
}
?>